### PR TITLE
Hue

### PR DESCRIPTION
**Pull request by abhay-agarwal**

### CODE REVIEW

**Comment by abhay-agarwal** [comment]
@jsancio can you please look over? I'll actually merge once the requisite changes have been made to HDFS.

**Comment by jsancio** [comment]
High-level note. Please make this package configurable.

**Comment by abhay-agarwal** [comment]
Okay. Which parameters should be configurable?

**Comment by abhay-agarwal** [comment]
updated @jsancio

**Comment by elingg** [comment]
Thanks @abhay-agarwal! Before merging we need to point the following URL here, https://github.com/mesosphere/universe/blob/master/repo/packages/H/hdfs/0/config.json#L11: https://s3.amazonaws.com/downloads.mesosphere.io/hdfs/artifacts/dcos/0.1.1/hdfs-mesos-0.1.1.tgz

**Comment by abhay-agarwal** [comment]
@jsancio This is good to ship!

File: repo/packages/H/hue/0/config.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "type": "object",
+  "properties": {
+    "hue/configuration": {
```
Comment (jsancio) [inline repo/packages/H/hue/0/config.json]: We now recommend people to use hierarchical options files. See Marathon and Cassandra. Thanks!

**Comment by elingg** [comment]
hi @abhay-agarwal, I just caught this, too. We also need to update the version to 0.1.1 here: https://github.com/mesosphere/universe/blob/master/repo/packages/H/hdfs/0/package.json#L3. With that change we also need to rebuild the index.

**Comment by abhay-agarwal** [comment]
Good to :ship: @jsancio

**Comment by abhay-agarwal** [comment]
I also updated the version number of HDFS and updated the HDFS subcommand as per Connor's comments.

**Comment by abhay-agarwal** [comment]
@jsancio Can I merge this?

File: repo/packages/H/hue/0/marathon.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                "hostPath": "/opt/mesosphere/etc/hadoop/",
+                "mode": "RO"
+            }
+        ]
+    },
+    "uris": [
+        "{{hue.configuration}}",
+        "{{httpfs.configuration}}"
+    ],
+    "cmd": "export PATH=/opt/mesosphere/bin:$PATH && \\cp -f /etc/hadoop/conf/*-site.xml /etc/hadoop-httpfs/conf/ && \\cp -f $MESOS_SANDBOX/httpfs-site.xml /etc/hadoop-httpfs/conf/ && service hadoop-httpfs start; \\cp -f $MESOS_SANDBOX/hue.ini /etc/hue/conf/hue.ini && /usr/lib/hue/build/env/bin/hue runserver 0.0.0.0:8000",
```
Comment (jsancio) [inline repo/packages/H/hue/0/marathon.json]: This is PATH manipulation looks very DCOS Image specific. Is this so we can get access to `service`? The user should be able to overwrite this to work on their env. cc @cmaloney

Comment (abhay-agarwal) [inline repo/packages/H/hue/0/marathon.json]: True. Let me fix.

**Comment by abhay-agarwal** [comment]
@jsancio I have updated this package. Take a look

File: repo/packages/H/hue/0/marathon.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                "containerPath": "/etc/hadoop/conf/",
+                "hostPath": "{{hadoop.conf-dir}}",
+                "mode": "RO"
+            }
+        ]
+    },
+    "uris": [
+        "{{hue.configuration}}",
+        "{{httpfs.configuration}}"
+    ],
```
Comment (elingg) [inline repo/packages/H/hue/0/marathon.json]: hi @abhay-agarwal, for this cmd, what you should do is directly download the Hadoop package that you need instead of mounting directories in DCOS image as volumes. We want to avoid dependencies on DCOS Image if possible, and it should be possible here. If you have questions about this, feel free to discuss with @cmaloney or ask me.

**Comment by elingg** [comment]
Thanks @abhay-agarwal, I took a look and left a comment that should be addressed. Once that's taken care of, we can do a final review.

**Comment by abhay-agarwal** [comment]
@elingg our system architecture is highly dependent on the hadoop that is actually at `/opt/mesosphere/bin/` because it references other configs. Similarly, I have to mount the entire `/opt/mesosphere/` folder because there are symlinks that are referenced throughout. Until we develop a better architecture for Hadoop, this is the best solution I can engineer.

**Comment by cmaloney** [comment]
I don't see a good reason why you couldn't download the hdfs, static config inside of the Hue app itself rather then relying on the copy of (more or less) the exact same in the host system. What is structurally so hard to download / add to path yourself?

**Comment by elingg** [comment]
hi @abhay-agarwal, +1 to @cmaloney's comment. Just download the hdfs tarball and use that instead of depending on DCOS Image. It has the config preset. That would be much cleaner.

**Comment by abhay-agarwal** [comment]
@elingg @cmaloney The only drawback there is that i'll have to update this package every time our HDFS tarball changes. Which could potentially be every week/month/etc.

**Comment by elingg** [comment]
@abhay-agarwal, that's okay, we just need to update the package to link to the latest released version of HDFS. We need to do this anyway in Universe each time there is a release.

**Comment by cmaloney** [comment]
If you use the base system ones:

1. You are getting a special exception from the rules (Don't depend on the host system), your service isn't self contained.
2. The system packages are going to change without warning / notice to your DCOS Service. This likely means at some point there will be a system update independent of your application update which would completely break Hue without notice to users. If things are self-contained you'll at least get "files exist" and hopefully some reasonable error messages about incompatible network APIs / protocol versions.

Practically this all depends on the hdfs-mesos framework, not the host HDFS for functionality. I think there is a way to specify that dependency.

**Comment by abhay-agarwal** [comment]
Agreed. I made the change in the packge so there's no more volume mounting. Would be great to have a dependency system in-place by the next time I have to make the update :shipit:

**Comment by elingg** [comment]
Yes, I think this is the best way to include this dependency is linking to it as opposed to using the base system. I will take a look at this PR ASAP.

File: repo/packages/H/hue/0/config.json [hunk 4]
```diff
@@ -0,0 +1,66 @@
+{
+  "type": "object",
+  "properties": {
+    "hdfs": {
+      "description": "Hadoop specific configuration properties",
+      "type": "object",
+      "properties": {
+        "scheduler-uri": {
```
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: nit: In description, it should say (must being with 'hdfs-mesos') instead of  "description": "HDFS package artifact tarball (must begin with 'hdfs')"

Comment (elingg) [inline repo/packages/H/hue/0/config.json]: Instead of calling this scheduler URI, this should be called package URI.

File: repo/packages/H/hue/0/config.json [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    "httpfs": {
+      "description": "Httpfs specific configuration properties",
+      "type": "object",
+      "properties": {
+        "configuration": {
+          "description": "A custom universally-accessible URL of a httpfs-site.xml file.",
+          "type": "string",
+          "default": "https://s3.amazonaws.com/downloads.mesosphere.io/hdfs/hue/httpfs-site.xml"
+        },
+        "port": {
```
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: The description needs to be updated here.

File: repo/packages/H/hue/0/config.json [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    "httpfs": {
+      "description": "Httpfs specific configuration properties",
+      "type": "object",
+      "properties": {
+        "configuration": {
+          "description": "A custom universally-accessible URL of a httpfs-site.xml file.",
+          "type": "string",
+          "default": "https://s3.amazonaws.com/downloads.mesosphere.io/hdfs/hue/httpfs-site.xml"
+        },
```
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: Quick question, why do we need port for "httpfs" and for "hue"?

Comment (abhay-agarwal) [inline repo/packages/H/hue/0/config.json]: This hue package bundles a httpfs server (its the only way to address an HA hdfs install). Currently, both are served via port 0, but if a user wants a specific port for either (to support some other tool, etc) they can.

Comment (elingg) [inline repo/packages/H/hue/0/config.json]: makes sense, thanks

File: repo/packages/H/hue/0/config.json [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        "scheduler-uri"
+      ]
+    },
+    "hue": {
+      "description": "Hue specific configuration properties",
+      "type": "object",
+      "properties": {
+        "configuration": {
+          "description": "A custom universally-accessible URL of a hue.ini file.",
+          "type": "string",
```
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: Let's move the bucket for Hue outside of HDFS. Remember that it will be used by Myriad, Spark, etc so it is a separate entity than HDFS.

Comment (abhay-agarwal) [inline repo/packages/H/hue/0/config.json]: I don't follow. Do you mean remove the tag in `package.json`?

File: repo/packages/H/hue/0/config.json (hunk 4 above)
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: nit: In description, it should say (must being with 'hdfs-mesos') instead of "description": "HDFS package artifact tarball (must begin with 'hdfs')"

File: repo/packages/H/hue/0/config.json (hunk 7 above)
Comment (elingg) [inline repo/packages/H/hue/0/config.json]: Let's make this "https://s3.amazonaws.com/downloads.mesosphere.io/hue/httpfs-site.xml" instead of "https://s3.amazonaws.com/downloads.mesosphere.io/hdfs/hue/httpfs-site.xml"

**Comment by abhay-agarwal** [comment]
@elingg updated with your changes.

**Comment by elingg** [comment]
LGTM! Any objections @jsancio or @cmaloney? If not, we should be ready to merge.

File: repo/packages/H/hdfs/0/command.json [hunk 8]
```diff
@@ -1,5 +1,5 @@
 {
   "pip": [
-    "git+https://github.com/mesosphere/dcos-hdfs.git@0.1.0#dcos-hdfs=0.1.0"
+    "git+https://github.com/mesosphere/dcos-hdfs.git@0.1.1#dcos-hdfs=0.1.1"
```
Comment (jsancio) [inline repo/packages/H/hdfs/0/command.json]: Do we have plans to remove this technical debt?

**Comment by cmaloney** [comment]
looks reasonable to me from a "doesn't touch the base image" perspective.

File: repo/packages/H/hdfs/0/command.json (hunk 8 above)
Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: I'm still not sure how we would send a local configuration file without encoding it.

Comment (jsancio) [inline repo/packages/H/hdfs/0/command.json]: I meant that we should be using wheels to package and install Python packages. We should be setting standard for others to follow.

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: +1 this would be nice to get rid of. I'm not familiar with the reasoning behind creating another repo, but I would like to remove this repo if we can. What course of action do you suggest @jsancio?

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: but if you mean rewriting the hdfs universe package to take different config parameters, its not on the roadmap yet.

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: The separate repo is how Universe packages install CLI tools. See how spark does it. If you remember from around a month ago, we had a push to 'make hdfs configurable' so I implemented the only solution that was within the time constraints. Since we have another three weeks now, we can add a better solution to the roadmap.

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: In the meantime, this update fixes https://mesosphere.atlassian.net/browse/DCOS-1093

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: Thanks for clarifying, @abhay-agarwal. So the separate repo is indeed necessary, but the way we access it in Universe is not ideal.  We can merge this now and then @abhay-agarwal could add a JIRA to fix this technical debt and work on it within the next 3 weeks. Does that sound like a good plan, @abhay-agarwal and @jsancio?

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: https://mesosphere.atlassian.net/browse/HDFS-85

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: Thanks. Actually I thought about it and configuring it would not be hard using options.json. We'd just ask the users to point to a URL with their config file, which is the common way things are handled in Universe. This would be a fairly quick fix, right, or am I missing something?

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: that fix is already available via 'scheduler-uris' in hdfs package. This is to allow a user to specify individual parameters in options.json.

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: Ok, discussed with others and for passing the config, the proper fix is https://mesosphere.atlassian.net/browse/HDFS-25. I closed HDFS-85 for this reason.

The main issue that @jsancio was referring to was a team city build and an automated update to an s3 bucket for the subcommand code: git+https://github.com/mesosphere/dcos-hdfs.git@0.1.1#dcos-hdfs=0.1.1. Since automated builds are important and hdfs itself is now built in team city, could you do this?

Comment (abhay-agarwal) [inline repo/packages/H/hdfs/0/command.json]: Can we move that to a separate PR? This is not an omnibus update to HDFS.

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: Sure, fine by me. Let's create a JIRA to automate the build of the dcos-hdfs (hdfs subcommand). Will you be looking into it? If so, assign it to yourself. If not, assign it to me for now.

Comment (elingg) [inline repo/packages/H/hdfs/0/command.json]: Cool, thanks for creating the JIRA and assigning to me. We should be good to go!

**Comment by elingg** [comment]
Ship It! :shipit:
